apps/wallet/src/ui/app/staking/stake/StakingCard.tsx

```diff
@@ -115,16 +151,41 @@ function StakingCard() {
             setSendError(null);
             try {
                 const bigIntAmount = parseAmount(amount, coinDecimals);
-                // TODO: add unstake functionality
-                if (unstake) return;
-                const response = await dispatch(
-                    stakeTokens({
-                        amount: bigIntAmount,
-                        tokenTypeArg: coinType,
-                        validatorAddress: validatorAddress,
-                    })
-                ).unwrap();
-                const txDigest = getTransactionDigest(response);
+                let response;
+                let txDigest;
+                if (unstake) {
+                    // check for delegation data
+                    if (
+                        !delegationData ||
+                        !stakeIdParams ||
+                        delegationData.delegation_status === 'Pending'
+                    ) {
+                        return;
+                    }
+                    response = await dispatch(
+                        unStakeToken({
+                            principalWithdrawAmount: bigIntAmount.toString(),
+                            delegation:
+                                delegationData.delegation_status.Active.id.id,
+                            stakedSuiId:
+                                delegationData.delegation_status.Active
+                                    .staked_sui_id,
+                        })
+                    );
+                    const payload =
+                        response.payload as SuiExecuteTransactionResponse;
+                    txDigest = getTransactionDigest(payload);
+                } else {
+                    response = await dispatch(
+                        stakeTokens({
```

Comment:
Probably could also easily move this to a `useMutation`

Reply:
done